compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp

```diff
@@ -405,19 +408,133 @@ CollapseInfo::getTransformedCollapsableLoops(AffineMap map) const {
   return transformedLoops;
 }
 
-// Update `collapsableLoops` by taking the set intersection with
-// `otherCollapsable` and update the reassociation indicies accordingly.
-bool CollapseInfo::updateCollapseViaIntersect(
-    const CollapsableLoopsSet &otherCollapsable) {
-  CollapsableLoopsSet toRemove;
-  for (auto elem : collapsableLoops) {
-    if (!otherCollapsable.contains(elem)) {
-      toRemove.insert(elem);
+FailureOr<SmallVector<ReassociationIndices>>
+CollapseInfo::getTransformedReassociation(AffineMap map) const {
+  if (!map) {
+    return failure();
+  }
+
+  SmallVector<ReassociationIndices> transformedReassociation(reassociation);
+  for (const auto &[idxOne, indicies] : llvm::enumerate(reassociation)) {
+    for (auto [idxTwo, elem] : llvm::enumerate(indicies)) {
+      assert(elem < map.getNumResults() && "index has no valid mapping");
+      auto dimExpr = dyn_cast<AffineDimExpr>(map.getResult(elem));
+      if (!dimExpr) {
+        continue;
+      }
+      transformedReassociation[idxOne][idxTwo] = dimExpr.getPosition();
     }
   }
-  collapsableLoops.set_subtract(toRemove);
-  updateReassociation();
-  return toRemove.size();
+  return transformedReassociation;
+}
+
+// Update CollapseInfo to ensure that all dimensions collapsable in `this` are
+// also collapsable in `otherInfo`. This means:
+// 1. Any dimension not collapsable in `otherInfo` should not be collapsable in
+// `this`
+// 2. For any pair of dimensions in `this`, if they are collapsable in
+// `consumerInfo`, they must be collapsable into the same dimension in
+// `consumerInfo` to be collapsable into the same dimension in `this`.
```

Comment:
nit: move the comment to where the function is declared.

Also, please document what the return value means.